thaw/src/auto_complete/mod.rs

```diff
@@ -82,39 +85,59 @@ pub fn AutoComplete(
         if let Some(on_select) = on_select {
             on_select.call(option_value);
         }
+        if allow_free_input {
+            select_option_index.set(None);
+        }
         is_show_menu.set(false);
     };
 
+    // we unset selection index whenever options get changed
+    // otherwise e.g. selection could move from one item to
+    // another staying on the same index
+    create_effect(move |_| {
+        options.get();
+        select_option_index.set(default_index);
+    });
+
     let on_keydown = move |event: ev::KeyboardEvent| {
         if !is_show_menu.get_untracked() {
             return;
         }
         let key = event.key();
         if key == *"ArrowDown" {
             select_option_index.update(|index| {
-                if *index == options.with_untracked(|options| options.len()) - 1 {
-                    *index = 0
+                if *index == Some(options.with_untracked(|options| options.len()) - 1) {
+                    *index = default_index
                 } else {
-                    *index += 1
+                    *index = Some(index.map_or(0, |index| index + 1))
                 }
             });
         } else if key == *"ArrowUp" {
             select_option_index.update(|index| {
-                if *index == 0 {
-                    *index = options.with_untracked(|options| options.len()) - 1;
+                if *index == Some(0) {
+                    if allow_free_input {
+                        *index = None
+                    } else {
+                        *index = Some(options.with_untracked(|options| options.len()) - 1);
+                    }
                 } else {
-                    *index -= 1
+                    *index = index.map(|index| index - 1)
```

Comment:
There's a marginal problem. When index is None, the last of the array should be selected.

Reply:
I'm not quite sure if that would be intuitive but I'll switch to that.

Reply:
Github's search box works the same way.